File: lib/sprite.rb
require 'drawable'
require 'color'
require 'tone'

# The sprite class.
# Sprites are the basic concept used to display
# characters, etc. on the game screen.
class Sprite < Drawable
  # Refers to the {Bitmap} used for the sprite's starting point.
  attr_reader :bitmap

  def bitmap=(bmp)
    @bitmap = bmp
    self.src_rect = Rect.new(0, 0, bmp.width, bmp.height)
  end

  # The box (Rect) taken from a bitmap.
  attr_accessor :src_rect

  # The sprite's visibility. If TRUE, the sprite is visible.
  attr_accessor :visible

  # The sprite's X-coordinate.
  attr_accessor :x

  # The sprite's Y-coordinate.
  attr_accessor :y

  # The viewport's Z-coordinate. The larger this value, the closer
  # to the player the viewport will be displayed. If multiple objects share
  # the same Z-coordinate, the more recently created object will be displayed
  # closest to the player.
  attr_accessor :z

  # The X-coordinate of the sprite's starting point.
  attr_accessor :ox

  # The Y-coordinate of the sprite's starting point.
  attr_accessor :oy

  # The sprite's X-axis zoom level. 1.0 denotes actual pixel size.
  attr_accessor :zoom_x

  # The sprite's Y-axis zoom level. 1.0 denotes actual pixel size.
  attr_accessor :zoom_y

  # The sprite's angle of rotation. Specifies up to 360 degrees of
  # counterclockwise rotation. However, drawing a rotated sprite is
  # time-consuming, so avoid overuse.
  attr_accessor :angle

  # Flag denoting the sprite has been flipped horizontally.
  # If TRUE, the sprite will be drawn flipped.
  attr_accessor :mirror

  # The Bush depth for that sprite. This is a pixel value denoting how much of
  # the sprite's lower portion will be displayed as semitransparent.
  # A simple way to convey a sense of a character's feet being obscured by
  # foliage and the like.
  attr_accessor :bush_depth

  # The sprite's opacity (0-255). Values out of range are automatically corrected.
  attr_accessor :opacity

  # The sprite's blending mode (0: normal, 1: addition, 2: subtraction).
  attr_accessor :blend_type

  # The color (Color) to be blended with the sprite.
  # Alpha values are used in the blending ratio.
  #
  # Handled separately from the color blended into a flash effect.
  # However, the color with the higher alpha value when displayed will
  # have the higher priority when blended.
  attr_accessor :color

  # The sprite's color {Tone}.
  attr_accessor :tone

  # Gets the {Viewport} specified when the tilemap was created.
  attr_reader :viewport

  def initialize(viewport = nil)
    super()
    @viewport = viewport
    @src_rect = Rect.new(0, 0, 0, 0)
    @visible = true
    @x = 0
    @y = 0
    @z = 0
    @ox = 0
    @oy = 0
    @zoom_x = 0
    @zoom_y = 0
    @angle = 0
    @mirror = false
    @bush_depth = 0
    @opacity = 255
    @blend_type = 0
    @color = Color.new(0, 0, 0, 0)
    @tone = Tone.new(0, 0, 0, 0)
  end

  # Frees the sprite. If the sprite has already been freed, does nothing.
  def dispose
    @disposed = true
  end

  # Returns TRUE if the sprite has been freed.
  def disposed?
    @disposed
  end

  # Begins flashing the sprite.
  # duration specifies the number of frames the flash will last.
  #
  # If color is set to nil, the sprite will disappear while flashing.
  def flash(color, duration)
  end

  # Refreshes the sprite flash. As a rule, this method is called once per frame.
  #
  # It is not necessary to call this method if no flash effect is needed.
  def update
  end

  def draw(bmp)
    unless bitmap.nil? || !visible || opacity == 0
      begin
        bmp.blt(x, y, bitmap, src_rect, opacity)
      rescue ArgumentError => e # WTF?
        #p x, y, bitmap, src_rect, opacity
      end
    end
  end
end

File: src/RMXPx/Sprite.cs
using System.Collections.Generic;

namespace RMXPx
{
    // TODO: throw when assigning blend_type less than 0 or greater than 2
    public class Sprite : IRenderable
    {
        private Bitmap _bitmap;
        public Bitmap Bitmap
        {
            get { return _bitmap; }
            set
            {
                _bitmap = value;
                SrcRect = _bitmap.GetRect();
            }
        }

        public Rect ActualSrcRect
        {
            get
            {
                if (Viewport == null)
                {
                    return SrcRect;
                }

                return new Rect(SrcRect.X + Viewport.OX + OX - X,
                                SrcRect.Y + Viewport.OY + OY - Y,
                                Viewport.Rect.Width,
                                Viewport.Rect.Height)
                                .Intersect(SrcRect);
            }
        }

        public bool Visible { get; set; }
        public Rect/*!*/ SrcRect { get; set; }
        public int X { get; set; }
        public int Y { get; set; }
        public int Z { get; set; }
        public int OX { get; set; }
        public int OY { get; set; }
        public int ZoomX { get; set; }
        public int ZoomY { get; set; }
        public int Angle { get; set; }
        public bool Mirror { get; set; }
        public int BushDepth { get; set; }
        public int Opacity { get; set; }
        public int BlendType { get; set; }
        public Color/*!*/ Color { get; set; }
        public Tone/*!*/ Tone { get; set; }
        private readonly Viewport _viewport;
        public Viewport Viewport
        {
            get { return _viewport; }
        }

        

        public bool Disposed { get; private set; }

        //public Sprite() : this(null)
        //{
        //}

        public Sprite(Viewport viewport) 
        {
            _viewport = viewport;
            if (_viewport != null)
            {
                _viewport.AddSprite(this);
                if (_viewport.Z == 102)
                {
                    
                }
            }

            SrcRect = new Rect(0, 0, 0, 0);
            Visible = true;
            X = 0;
            Y = 0;
            Z = 0;
            OX = 0;
            OY = 0;
            ZoomX = 0;
            ZoomY = 0;
            Angle = 0;
            Mirror = false;
            BushDepth = 0;
            Opacity = 255;
            BlendType = 0;
            Color = new Color(0, 0, 0, 0);
            Tone = new Tone(0, 0, 0, 0);
        }

        public void Dispose()
        {

        }   
    }
}